infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -128,13 +141,34 @@ public LogUnitServer(ServerContext serverContext) {
         dataCache = new LogUnitServerCache(config, streamLog);
         batchWriter = new BatchProcessor(streamLog, serverContext.getServerEpoch(), !config.isNoSync());
 
-        logCleaner = new StreamLogCompaction(streamLog, 10, 45, TimeUnit.MINUTES, ServerContext.SHUTDOWN_TIMER);
+        logCleaner = new StreamLogCompaction(
+                streamLog, resetLock, 10, 45, TimeUnit.MINUTES, ServerContext.SHUTDOWN_TIMER
+        );
     }
 
 
     @Override
     protected void processRequest(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
-        executor.submit(() -> getHandler().handle(msg, ctx, r));
+        executor.submit(
+                () -> {
+                    switch (msg.getMsgType()) {
+                        case READ_REQUEST:
+                        case KNOWN_ADDRESS_REQUEST:
+                        case COMPACT_REQUEST:
+                            if (msg.getEpoch() != serverContext.getServerEpoch()) {
+                                CorfuPayloadMsg<Long> payload = new CorfuPayloadMsg<>(
+                                        CorfuMsgType.WRONG_EPOCH,
+                                        serverContext.getServerEpoch()
+                                );
+                                r.sendResponse(ctx, msg, payload);
+                                return;
+                            }
```

Comment:
Could you elaborate why this is checked here? From my understanding, a message type like READ_REQUEST or KNOWN_ADDRESS_REQUEST already have their epochs checked in the method epochIsValid from IServerRouter, since their ignoreEpoch fields are set as false. What's the difference between the getServerEpoch provided by the router versus the one provided by the context?

Reply:
Oh I see, the epoch can become outdated as the request waits in the queue.